twirp/async_client.py

```diff
@@ -0,0 +1,69 @@
+import asyncio
+import json
+import aiohttp
+
+from . import exceptions
+from . import errors
+
+
+class AsyncTwirpClient:
+    def __init__(self, address, timeout=5, session=None):
+        self._address = address
+        self._timeout = timeout
+        self._session = session
+        self._should_close_session = False
+
+    def __del__(self):
+        if self._should_close_session:
+            try:
+                loop = asyncio.get_event_loop()
+                if loop.is_running():
+                    loop.create_task(self._session.close())
+                elif not loop.is_closed():
+                    loop.run_until_complete(self._session.close())
+            except RuntimeError:
+                pass
+
+    @property
+    def session(self):
+        if self._session is None:
+            self._session = aiohttp.ClientSession(
+                self._address, timeout=aiohttp.ClientTimeout(total=self._timeout)
+            )
+            self._should_close_session = True
+        return self._session
```

Comment:
The go implementation doesn't create a client automatically.

I think it should be the same here.

```suggestion
    def __init__(self, address: str, session: ClientSession) -> None:
        self._address = address
        self._session = session

    async def aclose(self) -> None:
        await self._session.close()
```

It can be used like this:
```py
http_client_options = {
    "timeout":  aiohttp.ClientTimeout(total=timeout_s)
}

async with aiohttp.ClientSession(**http_client_options) as session:
    client = AsyncHaberdasherClient(address, session)
    await client.MakeHat(...)
    ...

async with contextlib.aclosing(AsyncHaberdasherClient(
    address, 
    aiohttp.ClientSession(**http_client_options)
)) as client:
    await client.MakeHat(...)
    ...
```

Reply:
Great! Yes, I agree, and I didn't love the idea of creating the session automatically, but I wasn't sure if some would prefer a simpler way to use it without having to provide and manage the underlying session. (Because the existing sync implementation with `requests` has no such session requirement, I was hesitant to add one, and was trying to keep the usage similar for those who wanted to keep it that way.)

However:

- `aiohttp.ClientSession` can only be created within a coroutine, so if you are required to provide a session when you init the twirp client then you can only do so within a coroutine. (This is not how we currently init our client dependencies within our app, so I had planned to pass the client session with each request, but that might feel cumbersome to some folks)
- If the caller creates and manages the client session, I'd think they should probably handle the closing of it also, rather than having the twirp client close the session

Thoughts?

Reply:
I think you can solve your problem by turning the client into an async context manager.
It would be something like that:
```py
def init():
     client = AsyncHaberdasherClient(address, session=None)
...
async def use(client: AsyncHaberdasherClient):
    async with (
        aiohttp.ClientSession() as session,
        client.with_session(session),
    ):
        await client.MakeHat(...)
```
`with_session` can be implemented like this:
```python
@asynccontextmanager
async def with_session(session: ClientSession):
  backup, self._session = self._session, session
  yield
  backup, self._session = None, backup
```

Or using a factory? (need testing)
```py
def init():
     async def session_factory():
         return aiohttp.ClientSession()
     client = AsyncHaberdasherClient(address, session_factory=session_factory)
...
async def use(client: AsyncHaberdasherClient):
    async with client:
        await client.MakeHat(...)
```
You would need to implement `__aenter__` and `__aexit__` on `AsyncTwirpClient` to call the functions on the session.

Reply:
After thinking a bit more about this, maybe you could just pass the `http_client_options` instead of a factory to the constructor when creating the twirp client.

Then you can use `__aenter__` and `__aexit__` to create a temporary session which can be reused for several calls.
But it could also create a session directly inside `_make_request` on each call for simplicity when there is no need to reuse a client session.

Reply:
Thanks for the suggestions @MiLk! I don't mind passing the session in each request, but do you think I should add `with_session` or `session_factory` as something that's generally useful for others? I pushed some changes removing the auto-created client, and would be fine merging as-is but would like to cover other common use cases.

Reply:
I didn't see that the `session` could be passed directly with each request.
You can disregard my previous suggestions.
I think that's enough, and there is no need to make the code more complex.